pkg/monitortests/clusterversionoperator/legacycvomonitortests/operators.go

```diff
@@ -35,14 +35,94 @@ func testStableSystemOperatorStateTransitions(events monitorapi.Intervals) []*ju
 			}
 		}
 
-		return "We are not worried about Available=False or Degraded=True blips for stable-system tests yet.", nil
+		// For the non-upgrade case, if any operator has Available=False, fail the test.
+		if condition.Type == configv1.OperatorAvailable {
+			if condition.Status == configv1.ConditionFalse {
+				return "", nil
+			}
+		}
+		return "We are not worried about Degraded=True blips for stable-system tests yet.", nil
 	}
 
 	return testOperatorStateTransitions(events, []configv1.ClusterStatusConditionType{configv1.OperatorAvailable, configv1.OperatorDegraded}, except)
 }
 
+// isInUpgradeWindow determines if the given eventInterval falls within an upgrade window.
+// If we don't find any upgrade starting point, we assume the eventInterval is not in an upgrade window.
+// If we don't find any upgrade ending point, we assume the ending point is at the end of the test.
+func isInUpgradeWindow(eventList monitorapi.Intervals, eventInterval monitorapi.Interval) bool {
+	type upgradeWindowHolder struct {
+		startInterval *monitorapi.Interval
+		endInterval   *monitorapi.Interval
+	}
+
+	var upgradeWindows []*upgradeWindowHolder
+	var currentWindow *upgradeWindowHolder
+
+	// Scan through the event list to define all upgrade windows.
+	for _, event := range eventList {
+		if event.Source != monitorapi.SourceKubeEvent || event.Locator.Keys[monitorapi.LocatorClusterVersionKey] != "cluster" {
+			continue
+		}
+
+		reason := string(event.Message.Reason)
+		if reason == "UpgradeStarted" || reason == "UpgradeRollback" {
+
+			// We assume a rollback ends an upgrade window and starts a new one.
+			if reason == "UpgradeRollback" {
+				if currentWindow != nil && currentWindow.startInterval.Message.Reason == "UpgradeStarted" {
+					currentWindow.endInterval = &monitorapi.Interval{
+						Condition: monitorapi.Condition{
+							Message: monitorapi.Message{
+								Reason: event.Message.Reason,
+							},
+						},
+						From: event.From,
+						To:   event.To,
+					}
+				}
```

Comment:
I'm not sure if your assumption is correct, did you confirm that UpgradeRollback is not possible after UpgradeComplete? I would expect it to be but I don't know for sure. If so you'd want to handle currentWindow == nil here and open a new one?